### ISSUE TITLE
Rate Limits not working as specified

### ISSUE THREAD

**Issue by alecor191** [dapr/dapr#5604]
## In what area(s)?

/area runtime

## What version of Dapr?

CLI version: 1.9.1
Runtime version: 1.9.4

## Expected Behavior

Configuring rate limits for app `A` should result in `429` responses being returned when another app (`B`) sends requests at a frequency that exceeds the rate limit.

## Actual Behavior

There are 2 unexpected behaviors:

- When `B` sends requests to `A` at a frequency that exceeds `A`'s configured rate limit, there are no 429; all requests succeed
- When `A` sends requests to `B` (which hasn't _any_ rate limits configured), `A` receives 429 from `B`.

## Steps to Reproduce the Problem

I have 2 apps `A` and `B`. `A` is configured with a rate limit of 1 rps. `B` has no limits configured. Here the relevant yaml snippets:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: default-rate-limit
spec:
  type: middleware.http.ratelimit
  version: v1
  metadata:
    - name: maxRequestsPerSecond
      value: 1

# Config assigned to app A
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: A 
spec:
  ...
  httpPipeline:
    handlers:
    - name: default-rate-limit
      type: middleware.http.ratelimit

# Config assigned to app B
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: B 
spec:
  ...
  # nothing related to rate limits defined
```

The following results show what happened when running a `curl` command from a pod shell connected to either container `A` or `B`. Note that the two requests were sent back-to-back (<1 second apart), so that the second request should have exceeded the configured rate limit of 1 rps.

### Send requests from A -> A

```text
> curl -I -X GET -H "dapr-app-id: a" http://localhost:3500/info
HTTP/1.1 200 OK
...

> curl -I -X GET -H "dapr-app-id: a" http://localhost:3500/info
HTTP/1.1 429 Too Many Requests
...
```

**EXPECTED**: Sending a requests from the container hosting the rate-limited app `A` to itself, works as expected. The second request fails due to >1 rps request frequency.

### Send requests from B -> A

```text
> curl -I -X GET -H "dapr-app-id: a" http://localhost:3500/info
HTTP/1.1 200 OK
...

> curl -I -X GET -H "dapr-app-id: a" http://localhost:3500/info
HTTP/1.1 200 OK
...
```

**UNEXPECTED**: As shown, all requests succeed even though the second one should have failed with 429.

### Send requests from B -> B

```text
> curl -I -X GET -H "dapr-app-id: b" http://localhost:3500/info
HTTP/1.1 200 OK
...

> curl -I -X GET -H "dapr-app-id: b" http://localhost:3500/info
HTTP/1.1 200 OK
...
```

**EXPECTED**: When `B` sends requests with >1 rps to itself, all succeed as there are no rate limits configured for `B`.

### Send requests from A -> B

```text
> curl -I -X GET -H "dapr-app-id: b" http://localhost:3500/info
HTTP/1.1 200 OK
...

> curl -I -X GET -H "dapr-app-id: b" http://localhost:3500/info
HTTP/1.1 429 Too Many Requests
...
```

**UNEXPECTED**: There should be no 429 responses returned by `B`, as there are no rate limits configured for `B`. However, it seems sending requests from `A` to `B` results in 429 responses.

**Comment by yaron2** [comment]
This pipeline applies to calls made to the Dapr HTTP API, while I think what you're after is to apply rate limiting when Dapr sends requests to your app. In that case, try using appHttpPipeline as shown here: https://docs.dapr.io/developing-applications/middleware/#configuring-app-middleware-pipelines

**Comment by alecor191** [comment]
Thanks a lot @yaron2! I changed `httpPipeline` to `appHttpPipeline` and now it behaves as expected: all requsets to `A` are rate limited and none of the requests to `B`, regardless of what app sent the request.

Here is what I don't understand though:

- I followed the [Rate limiting guidelines](https://docs.dapr.io/reference/components-reference/supported-middleware/middleware-rate-limit/). They use `httpPipeline` to "_protect your application from denial of service (DOS) attacks_". Which is the use case I'm looking for.
- The doc you referenced mentions that "_Service-to-service invocation middleware components apply to all **OUTGOING** calls from Dapr sidecar_" with `appHttpPipeline`. For a rate limiting middleware I would have expected it to be applied on incoming traffic to the Dapr sidecar, not outgoing?

I do have one issue now with `appHttpPipeline`: It seems that it strips the path from the request URL before forwarding it to the destination app:

```text
> curl -X GET -H "dapr-app-id: a" http://localhost:3500/info

{
  "unmatchedPath": "/"
}
```

Notice how I pass `/info` path but what arrives on my app is `/`. As soon as I remove the `appHttpPipeline` it works again as expected. Is there anything else I would need to do to not have the URL path modified by the middleware?

**Comment by ItalyPaleAle** [comment]
> The doc you referenced mentions that "Service-to-service invocation middleware components apply to all OUTGOING calls from Dapr sidecar" with appHttpPipeline. For a rate limiting middleware I would have expected it to be applied on incoming traffic to the Dapr sidecar, not outgoing?

I can explain that part. The rate-limiting is applied when Dapr communicates with app B.

In this scenario, where app A calls app B:

```
App A --> Dapr A --> Dapr B -✔️-> App B
```

The middleware is applied when Dapr B calls App B (where I put the checkmark).

> I do have one issue now with appHttpPipeline: It seems that it strips the path from the request URL before forwarding it to the destination app:

This is a separate bug I can investigate.

**Comment by alecor191** [comment]
Thanks @ItalyPaleAle for the details, really helpful! Two questions:

- Should the [Rate limiting guidelines](https://docs.dapr.io/reference/components-reference/supported-middleware/middleware-rate-limit/) be updated then to show that we need to use `appHttpPipeline` instead of `httpPipeline`? I believe it would be useful to not have others run into the same issue as well.
- Would you like me to open a separate GitHub issue to track the path stripping issue?

**Comment by ItalyPaleAle** [comment]
> Should the [Rate limiting guidelines](https://docs.dapr.io/reference/components-reference/supported-middleware/middleware-rate-limit/) be updated then to show that we need to use appHttpPipeline instead of httpPipeline? I believe it would be useful to not have others run into the same issue.

I'd like to page @mcandeia for this as he wrote this and understands it best

> Would you like me to open a separate GitHub issue to track the path stripping issue?

Yes please :) If you can also add some repro steps/code, it would be very helpful. Does this happen only with the rate limiting middleware or with other ones too?

**Comment by alecor191** [comment]
@ItalyPaleAle sure, here the ticket tracking the path stripping issue: #5610

**Comment by ItalyPaleAle** [comment]
I believe this is fixed by dapr/components-contrib#2349 which will be released with Dapr 1.10
